proof/refine/ARM/Ipc_R.thy

```diff
@@ -3161,6 +3161,341 @@ lemma maybeReturnSc_corres:
   apply (clarsimp simp: obj_at'_def)
   done
 
+lemma thread_set_empty_tcb_sched_context_valid_tcbs[wp]:
+  "thread_set (tcb_sched_context_update Map.empty) t \<lbrace>valid_tcbs\<rbrace>"
+  by (wp thread_set_valid_tcbs) (simp add: valid_tcbs_def valid_tcb_def tcb_cap_cases_def)
+
+lemma threadSet_empty_tcbSchedContext_valid_tcbs'[wp]:
+  "threadSet (tcbSchedContext_update Map.empty) t \<lbrace>valid_tcbs'\<rbrace>"
+  by (wp threadSet_valid_tcbs') (simp add: valid_tcb'_def valid_tcbs'_def tcb_cte_cases_def)
+
+lemma update_valid_tcb[simp]:
+  "valid_tcb ptr tcb (s\<lparr>release_queue := q\<rparr>) = valid_tcb ptr tcb s"
```

Comment:
Isn't there some locale that provides these rewrites in Refine if you instantiate `valid_tcb` to it? I can't remember if we did these only for AInvs or not. (Something that says the predicate depends only on the heap). Same for `valid_tcbs'` below.

Reply:
I gave it a shot but could not immediately see how it should be done (I'm not expert with locales yet). I've added a `FIXME RT` instead.

Reply:
`valid_tcbs` is added since we started Refine, and it hasn't made its way into AInvs locale yet. In AInvs, we do have such a locale for `valid_objs`, which is very nice, so we should definitely incorporate `valid_tcbs` in there somehow at some point, but maybe not now.

Still, Ipc_R is probably not the right place for this, and there are a few `valid_tcbs` update listed in `TcbAcc_R` (where `valid_tcbs` is defined). So, can you move this there and clean things up a bit while you are here? There is VER1343 that talks about sorting out the `valid_objs/valid_tcbs` interaction, and it would be also good if we add something regarding locale there.